src/extra/components/xgm.py

```diff
@@ -417,13 +422,24 @@ def pulse_counts(self, sase=None):
                 key = f"pulseEnergy.numberOfSa{pg.value}BunchesActual"
             self._npulses_by_train[pg] = self._control_source[key].xarray()
 
+            slow_counts = self._npulses_by_train[pg]
+            pulse_energy = self.pulse_energy(sase=sase)
+            common_tids = np.intersect1d(pulse_energy.trainId, slow_counts.trainId)
+            fast_counts = np.count_nonzero(pulse_energy.sel(trainId=common_tids), axis=1)
+
+            counts_match = np.allclose(fast_counts, slow_counts.sel(trainId=common_tids))
+            if not counts_match:
+                warn(f"Slow data pulse counts ({key}) don't match the counts from fast data (data.intensityTD), data may be invalid!")
```

Comment:
If we know the slow data counts aren't always reliable and the difference from the fast data is a useful signal of that, does that imply the fast data are more reliable? Should we be using that instead?

Reply:
Yes, in my experience the fast data is more reliable and in some ways more useful than what we get from the pulses components because it will show the real number of pulses that are getting delivered. I don't know why but sometimes the number of pulses will change for... reasons... and for whatever reason that doesn't seem to be reflected in the bunch pattern table.

As for whether we should be using that instead of the slow data, I'm not sure. I am a little hesitant to make it do something other than the obvious 'just return whatever was saved', but as you say that's probably not very useful.

Reply:
> it will show the real number of pulses that are getting delivered. I don't know why but sometimes the number of pulses will change for... reasons... and for whatever reason that doesn't seem to be reflected in the bunch pattern table.

Hu, do you have a concrete example? That should not be possible, as they both are looking at exactly the same data - the XGM DOOCS server takes the bunch pattern to slice out the correct values from the train signal for `data.intensityTD`. That's the entire reason I don't trust it, it's an interpretation of the same signal through a 3rd party I don't control.

Reply:
Ok, so weirdly enough the `XrayPulses` component does report the right number but not `PumpProbePulses` :octopus: Example from p6156, r185 with the XGM:
![image](https://github.com/European-XFEL/EXtra/assets/5361518/dbe16139-1310-4e21-957d-bc9d9d6737b3)

And the pulses components:
![image](https://github.com/European-XFEL/EXtra/assets/5361518/1091aef5-7e2e-4755-af25-89d9146f970c)

My code was using `PumpProbePulses` so it was giving the wrong number of pulses. Is that a bug in `PumpProbePulses`?

Reply:
No, it's doing exactly what it is supposed to do:

<img width="604" alt="image" src="https://github.com/European-XFEL/EXtra/assets/56224768/67ce4ce6-44b4-4c0e-827a-db7e9bd7a397">


`PumpProbePulses` creates a unified pattern from both, i.e. it looks at every possible pulse the machine could have, and counts it as a filled pulse if **either** FEL, PPL **or both** are present.

Reply:
Ah ok, makes sense.